File: lib/obscurax/browser.ex
defmodule Obscurax.Browser do
  @moduledoc "Launches an obscura browser instance."

  @type t :: %__MODULE__{ref: reference()}

  defstruct [:ref]

  alias Obscurax.Nif

  @spec new(keyword()) :: {:ok, t()} | {:error, Obscurax.Error.t()}
  def new(opts \\ []) do
    opts_map = %{
      stealth: Keyword.get(opts, :stealth, false),
      proxy: Keyword.get(opts, :proxy),
      user_agent: Keyword.get(opts, :user_agent),
      storage_dir: Keyword.get(opts, :storage_dir)
    }

    case Nif.browser_new(opts_map) do
      {:ok, ref} -> {:ok, %__MODULE__{ref: ref}}
      {:error, error} -> {:error, error}
    end
  end

  @spec new!(keyword()) :: t()
  def new!(opts \\ []) do
    case new(opts) do
      {:ok, browser} -> browser
      {:error, error} -> raise error
    end
  end

  @spec new_page(t()) :: {:ok, Obscurax.Page.t()} | {:error, Obscurax.Error.t()}
  def new_page(%__MODULE__{ref: ref}) do
    case Nif.browser_new_page(ref, self()) do
      {:ok, page_ref} -> {:ok, %Obscurax.Page{ref: page_ref}}
      {:error, error} -> {:error, error}
    end
  end

  @spec new_page!(t()) :: Obscurax.Page.t()
  def new_page!(browser) do
    case new_page(browser) do
      {:ok, page} -> page
      {:error, error} -> raise error
    end
  end

  @spec cookies(t()) :: Obscurax.CookieStore.t()
  def cookies(%__MODULE__{ref: ref}) do
    {:ok, store_ref} = Nif.browser_cookies(ref)
    %Obscurax.CookieStore{ref: store_ref}
  end
end

File: native/obscurax/src/browser.rs
use std::sync::Arc;

use obscura::Browser;
use rustler::{Encoder, Env, NifResult, Resource, ResourceArc, Term};

use crate::atoms;
use crate::error::nif_error;
use crate::page_thread::spawn_page_thread;

pub struct BrowserHandle {
    pub browser: Arc<Browser>,
}

#[rustler::resource_impl]
impl Resource for BrowserHandle {}

fn map_get_bool(term: Term, key: &str) -> bool {
    term.map_get(key)
        .ok()
        .and_then(|t| t.decode::<bool>().ok())
        .unwrap_or(false)
}

fn map_get_string(term: Term, key: &str) -> Option<String> {
    let val = term.map_get(key).ok()?;
    if val.is_atom() {
        let atom: rustler::Atom = val.decode().ok()?;
        if atom == rustler::types::atom::nil() {
            return None;
        }
    }
    val.decode::<String>().ok()
}

#[rustler::nif]
pub fn browser_new<'a>(env: Env<'a>, opts: Term<'a>) -> NifResult<Term<'a>> {
    let stealth = map_get_bool(opts, "stealth");
    let proxy = map_get_string(opts, "proxy");
    let user_agent = map_get_string(opts, "user_agent");
    let storage_dir = map_get_string(opts, "storage_dir");

    let mut config = obscura::BrowserConfig::default();
    config.stealth = stealth;
    config.proxy = proxy;
    config.user_agent = user_agent;
    if let Some(dir) = storage_dir {
        config.storage_dir = Some(std::path::PathBuf::from(dir));
    }

    match Browser::build(config) {
        Ok(browser) => {
            let handle = ResourceArc::new(BrowserHandle {
                browser: Arc::new(browser),
            });
            Ok((atoms::ok(), handle).encode(env))
        }
        Err(e) => Err(rustler::Error::Term(Box::new(nif_error(
            "internal",
            e.to_string(),
        )))),
    }
}

#[rustler::nif]
pub fn browser_new_page<'a>(
    env: Env<'a>,
    handle: ResourceArc<BrowserHandle>,
    pid: rustler::LocalPid,
) -> NifResult<Term<'a>> {
    match spawn_page_thread(handle.browser.clone(), pid) {
        Ok(page) => {
            let arc = ResourceArc::new(page);
            Ok((atoms::ok(), arc).encode(env))
        }
        Err(e) => Err(rustler::Error::Term(Box::new(e))),
    }
}

#[rustler::nif]
pub fn browser_cookies<'a>(
    env: Env<'a>,
    handle: ResourceArc<BrowserHandle>,
) -> NifResult<Term<'a>> {
    let store = handle.browser.cookies();
    let arc = ResourceArc::new(crate::cookie::CookieStoreHandle::new(store));
    Ok((atoms::ok(), arc).encode(env))
}
